### PR TITLE
Add a Network Tunnel Activity Class

### PR DESCRIPTION
**Pull request by mikeradka**
#### Related Issue: 
#908 (Create a Network Tunnel Activity Class)

#### Description of changes:

This PR creates a new class in the `Network` category to allow for OCSF support of `Network Tunnel` events spanning data sources including but not limited to Cisco ASA/FTD, Palo Alto Globalprotect, and other products that support network tunneling.

image

### CODE REVIEW

File: dictionary.json [hunk 1]
```diff
@@ -843,6 +843,11 @@
       "is_array": true,
       "type": "string_t"
     },
+    "client": {
+      "caption": "Client",
+      "description": "The client endpoint. See specific usage.",
+      "type": "network_endpoint"
+    },
```
[the remark this answers is not in the document]
Comment (mikeradka) [reply dictionary.json:850]: I struggled with it a little bit, but the issue I encounter is that `src` and `dst` apply only where there is directionality.

In the context of a tunnel, this represents, `the remote host for the tunnel. This is usually associated with the public (WAN) ip of the tunnel. This is where the data is forwarded and undergoes decapsulation.`

I also considered naming this `remote` or `remote_endpoint` (i prefer less words if possible)

File: dictionary.json [hunk 2]
```diff
@@ -2119,6 +2124,11 @@
       "enum": {},
       "type": "integer_t"
     },
+    "interface": {
+      "caption": "Interface",
+      "description": "The interface object. See specific usage.",
+      "type": "network_interface"
+    },
```
Comment (Aniak5) [inline dictionary.json:2131]: This makes me wonder if we should rename the `network_interface` object to `interface`. Though every attribute underneath the `network_interface` object says `... network interface...`. Maybe it would be better to just use the `network_interface` object and not add this new attribute altogether. We may want to use `interface` attribute for something not related to network in the future.

Comment (mikeradka) [reply dictionary.json:2131]: I thought about that as well. `network_interface` makes querying a bit more clunky.

For instance:

- `tunnel.network_interface.ip` versus `tunnel.interface.ip`

File: dictionary.json [hunk 3]
```diff
@@ -3108,6 +3118,11 @@
       "description": "The unique identifier of a Cloud project.",
       "type": "string_t"
     },
+    "protocol_id": {
+      "caption": "Protocol ID",
+      "description": "The normalized networking protocol ID. See specific usage.",
+      "type": "integer_t"
+    },
```
Comment (Aniak5) [inline dictionary.json:3125]: Wondering if we should replace `protocol_num` with this. That would be in another PR and a breaking change but worth thinking about.

File: objects/tunnel.json [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  "caption": "Tunnel",
+  "name": "tunnel",
+  "description": "The tunnel object.",
+  "extends": "object",
+  "attributes": {
+    "client": {
+      "description": "The information about the client host (remote endpoint) for the tunnel. This is usually associated with the public (WAN) ip of the tunnel. This is where the data is forwarded and undergoes decapsulation.",
+      "group": "primary",
+      "requirement": "recommended"
+    },
```
Comment (Aniak5) [inline objects/tunnel.json:11]: Are there any instances where this differs from one or both of the other endpoints (`src_endpoint`, `dst_endpoint`)? Maybe if we have unidirectional flows? If it always matches one of these then this seems like redundant information thats just providing directionality information.

File: dictionary.json (hunk 3 above)
Comment (mikeradka) [reply dictionary.json:3125]: I think they are two different things.

`protocol_num` is `The TCP/IP protocol number, as defined by the Internet Assigned Numbers Authority (IANA). Use -1 if the protocol is not defined by IANA. See [Protocol Numbers](https://www.iana.org/assignments/protocol-numbers/protocol-numbers.xhtml). For example: 6 for TCP and 17 for UDP.` (the link makes it more clear)

Whereas `protocol_id` is an enum for protocols which apply to a specific networking context - in this case, the tunneling protocols.

Comment (Aniak5) [reply dictionary.json:3125]: Yeah they definitely are. For OCSF consistency it would make sense to label it `protocol_id` and then overwrite the ids with the TCP/IP protocol number, kind of like what we did with `rcode_id`. But from people familiar with IANA `protocol_num` makes more sense.

File: dictionary.json (hunk 2 above)
Comment (Aniak5) [reply dictionary.json:2131]: Yeah its definitely less clunky and more manageable. It seems like its used in the `DHCP` event class but under a new attribute of `relay` and then within the `network_interfaces` array. We might be able to rename it without breaking anything!?

File: dictionary.json (hunk 3 above)
Comment (mikeradka) [reply dictionary.json:3125]: I didn't realize that was actually possible :) Let me give it a try!

File: dictionary.json (hunk 2 above)
Comment (mikeradka) [reply dictionary.json:2131]: I *think* that might be okay, but maybe it should be a separate PR just so we can undergo a full due-diligence on it. Thoughts?

File: dictionary.json (hunk 3 above)
Comment (mikeradka) [reply dictionary.json:3125]: I think so, but I would opt for us to tackle that in a separate PR just for  the sake of due-diligence.

File: dictionary.json (hunk 2 above)
Comment (floydtree) [reply dictionary.json:2131]: I can see it can be a little inconvenient `tunnel.network_interface.ip` is the end column to be queried vs `tunnel.interface.ip`. 
Do you also then want to create an `interfaces` attribute to replace `network_interfaces`? If we don't do that, then I am concerned about the scope for confusion within the framework where the same attribute is being referred to with wholly different names (`interface`, `network_Interfaces`) with only difference being one is a simple object, other is an array of that same object. We can go the deprecation path for `network_interfaces` and replace it with `interfaces` if there's a major benefit in doing that, but I feel we should be very vigilant about deprecation and eventual breaking changes that users, consumers, producers will have to deal with.
I am currently not convinced either way, but I am leaning towards not creating interface and use network_interface instead to remain consistent and avoid eventual breaking changes.

Comment (mikeradka) [reply dictionary.json:2131]: @floydtree @Aniak5  I am thinking there might be a way to resolve this as well as the above concern about `client`:

- Change `client` to `remote` or `remote_endpoint` (i prefer less words)
- Similarly, create a `local` or `local_endpoint` attribute of type `network_endpoint`, and get rid of this `interface` attribute. Tunnel interface can then be captured via `local_endpoint.interface_name` and IP in `local_endpoint.ip`

Thoughts?

File: dictionary.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (cmcginley-splunk) [reply dictionary.json:850]: I also have mixed feelings about the client. I like the extraction of the tunnel info into a separate object; but it feels a little confusing that the client interface and tunnel interface are now referenced as:

* `tunnel.client.ip`
* `tunnel.interface.ip`

There was an immediate clarity to the original `client_interface` and `tunnel_interface` that we had which the above abstracts away

File: events/network/tunnel_activity.json [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        },
+        "2": {
+          "caption": "Close",
+          "description": "Close a tunnel."
+        },
+        "3": {
+          "caption": "Renew",
+          "description": "Renew a tunnel."
+        }
+      },
```
Comment (pagbabian-splunk) [inline events/network/tunnel_activity.json:23]: Missing the standard `0` - `Unknown` and `99` - `Other` for the standard sibling `activity_name`.

**Comment by mikeradka** [comment]
Flattened the class per valuable feedback from mappers.

File: dictionary.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (mikeradka) [reply dictionary.json:850]: @cmcginley-splunk I flattened the class per offline discussions.

File: events/network/tunnel_activity.json (hunk 5 above)
Comment (mikeradka) [reply events/network/tunnel_activity.json:23]: For some reason, the `Unknown` and `Other` work in the browser without having to explicitly include them:

image

Is this desired, or should we still manually add them?

File: dictionary.json (hunk 2 above)
Comment (mikeradka) [reply dictionary.json]: Closing the loop - as concluded in today's Network Sync, replacing `client` with `src_endpoint`:

image

File: objects/tunnel.json (hunk 4 above)
Comment (mikeradka) [reply objects/tunnel.json]: Closing the loop - as concluded in today's Network sync, replacing `client` with `src_endpoint` and adding `dst_endpoint` as defined in terms of the initiator/listener of the connection:

image

image

File: dictionary.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (mikeradka) [reply dictionary.json]: Closing the loop - as concluded in today's Network workstream, replacing `client` with `src_endpoint`, adding `dst_endpoint`, and defining the attributes with regards to the initiator/listener of the tunnel connection, respecitively:

image

image

**Review by floydtree — APPROVED** [review]

**Review by zschmerber — APPROVED** [review]

**Comment by Aniak5** [comment]
This looks great @mikeradka! Thanks for creating this class. Just curious, is there a reason that we aren't extending the base network class? it seems like the additional attributes of `tls`, `proxy` and `app_name` could be applicable . Or were we trying to avoid all the network profiles?

**Review by pagbabian-splunk — APPROVED** [review]

**Review by pagbabian-splunk — APPROVED** [review]

**Review by Aniak5 — APPROVED** [review]
Looks awesome!

**Comment by mikeradka** [comment]
> This looks great @mikeradka! Thanks for creating this class. Just curious, is there a reason that we aren't extending the base network class? it seems like the additional attributes of `tls`, `proxy` and `app_name` could be applicable . Or were we trying to avoid all the network profiles?

It was really just for simplicity, but after we dug into it a bit further it looks like extending `network` gives the capability of adding some useful extra context. Thanks for suggesting! Change is in.

**Review by pagbabian-splunk — APPROVED** [review]

**Review by zschmerber — APPROVED** [review]
